src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,165 @@
+            current = current.next;
+        }
+        newNode.next = current;
+        newNode.prev = current.prev;
+        if (current.prev != null) {
+            current.prev.next = newNode;
+        } else {
+            head = newNode;
+        }
+        current.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null) {
+            throw new NullPointerException("List cannot be null");
+        }
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current.data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        T oldValue = current.data;
+        current.data = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        if (current.prev != null) {
+            current.prev.next = current.next;
+        } else {
+            head = current.next;
+        }
+        if (current.next != null) {
+            current.next.prev = current.prev;
+        } else {
+            tail = current.prev;
+        }
+        size--;
+        return current.data;
     }
-
     @Override
     public boolean remove(T object) {
-        return false;
+        if (object == null) {
+            Node<T> current = head;
+            while (current != null) {
+                if (current.data == null) {
+                    if (current.prev != null) {
+                        current.prev.next = current.next;
+                    } else {
+                        head = current.next;
+                    }
+                    if (current.next != null) {
+                        current.next.prev = current.prev;
+                    } else {
+                        tail = current.prev;
+                    }
+                    size--;
+                    return true;
+                }
+                current = current.next;
+            }
+            return false;
+        } else {
+            Node<T> current = head;
+            while (current != null) {
```

Comment:
Optimize the node iteration by iterating only the first or second half of the list based on the index value to improve performance.